### PR TITLE
Send SIGKILL to containers if send SIGTERM failed when docker daemon exit

### PR DESCRIPTION
**Pull request by hustcat**
Send SIGKILL to containers if send SIGTERM failed or timeout when docker daemon exit. Prevent the daemon to wait forever if process in container ingore SIGTERM signal.

Signed-off-by: Ye Yin eyniy@qq.com

### CODE REVIEW

**Comment by icecrime** [comment]
Ping @tiborvass @vieux!

**Comment by tiborvass** [comment]
@hustcat why don't we just reuse `container.Stop(10)` ? I see that most of the logic is redundant.

**Comment by hustcat** [comment]
@tiborvass Thanks. container.Stop(10) is better, I have changed it.

File: daemon/daemon.go [hunk 1]
```diff
@@ -1053,10 +1053,9 @@ func (daemon *Daemon) shutdown() error {
 
 			go func() {
 				defer group.Done()
-				if err := c.KillSig(15); err != nil {
-					logrus.Debugf("kill 15 error for %s - %s", c.ID, err)
+				if err := c.Stop(10); err != nil {
+					logrus.Infof("Stop container %s error - %s", c.ID, err)
```
Comment (tiborvass) [inline daemon/daemon.go]: @hustcat would you mind changing `%s` to `%v` for `err` ?

**Comment by tiborvass** [comment]
@crosbymichael do you agree with this PR?

File: daemon/daemon.go (hunk 1 above)
Comment (hustcat) [inline daemon/daemon.go]: Of course:)

**Comment by crosbymichael** [comment]
Let me think real quick.  I think we did this for a reason for some of the shutdown usecases so that the container's would keep running without the daemon and we would kill and restart at docker boot

**Comment by erikh** [comment]
mysql in a container + SIGKILL == bad time

**Comment by unclejack** [comment]
@crosbymichael is right, this was done a while ago so that we didn't kill unresponsive containers on shutdown. @erikh is also right about SIGKILL being really bad for databases. Stopping your Docker daemon and SIGKILL-ing a database doing maintenance or processing important data would be very bad.

The right solution would be to:
1. have a proper init with signal handling running in the containers
2. run containers without always requiring a restart when a restart of the Docker daemon is required (e.g. individually supervised containers)
3. hot upgrade of the supervising element from Docker when Docker is upgraded

https://github.com/unclejack/image-checker is something I've written to help find images which don't respond to SIGTERM. Something like that should be used to check if the images are OK.

**Comment by icecrime** [comment]
Sorry, I'm closing this for the previously mentioned reasons. Please feel free to comment here and ask us to reopen if you have a strong case for it!

**Pull request closed without merging**